lib/dfe/analytics/railtie.rb

```diff
@@ -11,6 +11,9 @@ class Railtie < Rails::Railtie
 
       initializer 'dfe.analytics.insert_middleware' do |app|
         app.config.middleware.use DfE::Analytics::Middleware::RequestIdentity
+
+        app.config.middleware.insert_before \
+          ActionDispatch::Static, DfE::Analytics::Middleware::SendCachedPageRequestEvent
```

Comment:
I believe it will just at to the end of the stack, like `use` does. But Rails should always have this present. This is what our friend Chat GPT says:

In Rails, the `ActionDispatch::Static` middleware is included by default in the middleware stack for new applications. This middleware is responsible for serving static files from the `public` directory of your Rails application.

However, it's important to note that the exact middleware stack can be customized in the `config/application.rb` file of your Rails application. So, it's possible to remove or reorder the `ActionDispatch::Static` middleware in the stack if needed.

That being said, it's generally not recommended to remove the `ActionDispatch::Static` middleware, as it's an essential part of serving static assets in a Rails application.

Reply:
💟 ChatGPT

Reply:
OK seems reasonable — so it would only fail to run if ever someone used something other than `ActionDispatch::Static` as the caching layer. Maybe worth one line in the docs?

Reply:
Done

Reply:
I think this change is causing issues when precompiling the assets, specifically after updating `dfe-analytics` in our app we're starting to see errors when trying to deploy.

https://github.com/DFE-Digital/apply-for-qualified-teacher-status/actions/runs/4609526294/jobs/8146775822

```sh
 > [builder 11/12] RUN RAILS_ENV=production     SECRET_KEY_BASE=required-to-run-but-not-used     GOVUK_NOTIFY_API_KEY=required-to-run-but-not-used     REDIS_URL=redis://required-to-run-but-not-used     bundle exec rails assets:precompile:
#18 3.549 rails aborted!
#18 3.549 No such middleware to insert before: ActionDispatch::Static
#18 3.549 /app/config/environment.rb:5:in `<main>'
```

It sounds like perhaps `ActionDispatch::Static` isn't added to the middleware stack when precompiling the assets - I'm not entirely clear on how the middleware stack interacts with the asset precompiling.

We'll revert this change for now in Apply for QTS, but I wonder if there's a way of only inserting the `DfE::Analytics::Middleware::SendCachedPageRequestEvent` middleware if the `ActionDispatch::Static` middleware exists?

Reply:
I can confirm we had to [revert the upgrade to 1.8.0 on Refer](https://github.com/DFE-Digital/refer-serious-misconduct/pull/681) because of an error:

```
bundle exec rails assets:precompile
#19 3.196 rails aborted!
#19 3.196 No such middleware to insert before: ActionDispatch::Static
```

https://github.com/DFE-Digital/refer-serious-misconduct/actions/runs/4730348588/jobs/8394175837

Reply:
Sorry chaps — fix on way!

Reply:
Ticket opened on DI Trello board:

https://trello.com/c/fBBPetwX/1107-refer-serious-misconduct-apply-for-qts-unable-to-upgrade-to-latests-dfe-analytics-gem